File: general_algorithm/backtracking/0079_word_search.js
/*
Given an m x n grid of characters board and a string word, return true if 
word exists in the grid.

The word can be constructed from letters of sequentially adjacent cells, 
where adjacent cells are horizontally or vertically neighboring. 
The same letter cell may not be used more than once.
*/

// https://leetcode.com/problems/word-search/description/

var exist = function (board, word) {
  let m = board.length,
    n = board[0].length;
  var dfs = function (i, j, index, visited) {
    if (
      i < 0 ||
      j < 0 ||
      i >= m ||
      j >= n ||
      board[i][j] !== word[index] ||
      visited[i][j]
    )
      return false;
    if (index === word.length - 1) return true;
    visited[i][j] = true;
    let res =
      dfs(i + 1, j, index + 1, visited) ||
      dfs(i - 1, j, index + 1, visited) ||
      dfs(i, j + 1, index + 1, visited) ||
      dfs(i, j - 1, index + 1, visited);
    visited[i][j] = false;
    return res;
  };
  visited = new Array(m).fill(0).map(() => new Array(n).fill(false));
  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      if (dfs(i, j, 0, visited)) return true;
    }
  }
  return false;
};

File: general_algorithm/backtracking/0079_word_search.py
# Given an m x n grid of characters board and a string word, return true if word exists in the grid.

# The word can be constructed from letters of sequentially adjacent cells, where adjacent cells are horizontally or vertically neighboring. The same letter cell may not be used more than once.

# https://leetcode.com/problems/word-search/description/

# backtracking
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m, n, l = len(board), len(board[0]), len(word)
        
        def dfs(i, j, k):
            # base case
            if not 0 <= i < m or not 0 <= j < n or board[i][j] != word[k]: 
                return False
            if k == l - 1: 
                return True
            # recursive case
            board[i][j] = ''
            res = dfs(i + 1, j, k + 1) or dfs(i - 1, j, k + 1) or dfs(i, j + 1, k + 1) or dfs(i, j - 1, k + 1)
            board[i][j] = word[k]
            return res

        for i in range(m):
            for j in range(n):
                if dfs(i, j, 0): return True
        return False
        

